rpc: add circuit breaker to connection dialing

petermattis opened the pull request:
The circuit breakers are powered by heartbeat success and failures.

Adjust the RaftTransport circuit breakers to only throttle node address
resolution (which is all they were throttling previously).

Fix kv.grpcTransportFactory to not error when the breaker to one of the
replicas is open.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8292)

<!-- Reviewable:end -->

petermattis committed:
rpc: add circuit breaker to connection dialing

The circuit breakers are powered by heartbeat success and failures.

Adjust the RaftTransport circuit breakers to only throttle node address
resolution (which is all they were throttling previously).

Fix kv.grpcTransportFactory to not error when the breaker to one of the
replicas is open.

Fixes #8290.

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[kv/transport.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOG2Ociy5aOP8SfZDOZ:-KOG2Ociy5aOP8SfZDO_:1943041577) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f45cdaf42f110123ebbff18a72075dfb3b2a76/kv/transport.go#L126)):*
> ```Go
> 				continue
> 			}
> 			return nil, err
> ```

Returning an error here seems wrong even if this can never happen. Seems better to only return an error from `grpcTransportFactory` if we couldn't get a client to any replica.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender.go, line 921 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGebKbike4gUUY3oFR:-KOGebKbike4gUUY3oFS:-434500904) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/kv/dist_sender.go#L921)):*
> ```Go
> 	}
> 	defer transport.Close()
> 	if transport.IsExhausted() {
> ```

how come this was needed?

---

*[kv/transport.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOG2Ociy5aOP8SfZDOZ:-KOGdqq2GDq6BuBjPdmz:-1083311584) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f45cdaf42f110123ebbff18a72075dfb3b2a76/kv/transport.go#L126)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Returning an error here seems wrong even if this can never happen. Seems better to only return an error from `grpcTransportFactory` if we couldn't get a client to any replica.
</details>
See also https://github.com/cockroachdb/cockroach/issues/7494.

By the way, why can't this happen? `grpc.Dial` can still return errors.

---

*[rpc/context.go, line 167 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGdwjpE22HwF5KGxtI:-KOGdwjpE22HwF5KGxtJ:-613578362) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L167)):*
> ```Go
> 
> 	if ctx.conns.breakers == nil {
> 		ctx.conns.breakers = make(map[string]*circuit.Breaker)
> ```

can this be done eagerly instead?

---

*[rpc/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeCXWwcPyUONBDUgk:-KOGeCXWwcPyUONBDUgl:-1286775909) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L252)):*
> ```Go
> }
> 
> // GetBreaker returns the circuit breaker for the specified remote address.
> ```

this is only used in tests. remove?

---

*[rpc/context_test.go, line 418 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeS-L3UpREJRtCkKY:-KOGeS-MgxZh26x4kj1I:-1664957194) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context_test.go#L418)):*
> ```Go
> 			return nil
> 		}
> 		return errors.Errorf("breaker not open")
> ```

`errors.Wrapf("expected open breaker", err)`

---

*[storage/raft_transport.go, line 233 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGdUk3C9EOzmRBytfA:-KOGdUk3C9EOzmRBytfB:2102165218) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/storage/raft_transport.go#L233)):*
> ```Go
> 	conn, err := t.rpcContext.GRPCDial(addr.String())
> 	if err != nil {
> 		return nil
> ```

maybe do something with the error?

---

*[storage/raft_transport.go, line 239 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGd_5pY5VJGLVp-g5x:-KOGd_5pY5VJGLVp-g5y:-46764404) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/storage/raft_transport.go#L239)):*
> ```Go
> 
> // processQueue opens a Raft client stream and sends messages from the
> // designated queue (ch) via that client, exiting when an error is received or
> ```

s/client/stream/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender.go, line 921 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGebKbike4gUUY3oFR:-KOGkOycpn1Rrdk1jxys:-983385090) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/kv/dist_sender.go#L921)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this was needed?
</details>
We can't call `transport.Send` if `IsExhausted` returns false. Previously, we would have gotten an error from `transportFactory`, but now we'll get an "empty" transport.

---

*[kv/transport.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOG2Ociy5aOP8SfZDOZ:-KOGkjg1kYYq_DqfY7Qn:1412091300) ([raw file](https://github.com/cockroachdb/cockroach/blob/26f45cdaf42f110123ebbff18a72075dfb3b2a76/kv/transport.go#L126)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> See also https://github.com/cockroachdb/cockroach/issues/7494.
> 
> By the way, why can't this happen? `grpc.Dial` can still return errors.
</details>
True, it can return errors, but it almost never does. Yeah, we should probably hook into `grpc.Balancer`, but that's a much more extensive change.

---

*[rpc/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeCXWwcPyUONBDUgk:-KOGlNJTqjL5BTFFXsrQ:476516510) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is only used in tests. remove?
</details>
I've unexported. That's what you intended, right?

---

*[rpc/context_test.go, line 418 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeS-L3UpREJRtCkKY:-KOGlaleaF65MVtX-Dqj:-1967605703) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context_test.go#L418)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `errors.Wrapf("expected open breaker", err)`
</details>
`errors.Wrapf` returns nil if the supplied `err` is nil. We need to return a non-nil error.

---

*[storage/raft_transport.go, line 233 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGdUk3C9EOzmRBytfA:-KOGmLeckl0_tPXi5O9q:-950384414) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/storage/raft_transport.go#L233)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> maybe do something with the error?
</details>
Logged it if it isn't `ErrBreakerOpen`.

---

*[storage/raft_transport.go, line 239 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGd_5pY5VJGLVp-g5x:-KOGmLEB9XkTeUZsLj35:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/storage/raft_transport.go#L239)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/client/stream/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[rpc/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeCXWwcPyUONBDUgk:-KOGtLKW2wK05Dz5hkoE:-1491013546) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L252)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I've unexported. That's what you intended, right?
</details>
Eh, sure. I'd remove it entirely, it has one caller.

---

*[rpc/context_test.go, line 418 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeS-L3UpREJRtCkKY:-KOGtOEqhj0RHlsCeWlL:-1228832654) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context_test.go#L418)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `errors.Wrapf` returns nil if the supplied `err` is nil. We need to return a non-nil error.
</details>
ah, okay. interpolate the error in then - what I was getting at is that we'll want to see the error that came out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-:-KOGtkVoPVjlphjBGg8e:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[rpc/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeCXWwcPyUONBDUgk:-KOGvRl-os4Jo9iJeN7E:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Eh, sure. I'd remove it entirely, it has one caller.
</details>
Done.

---

*[rpc/context_test.go, line 418 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGeS-L3UpREJRtCkKY:-KOGvTLjXrKFACckSSW9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context_test.go#L418)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ah, okay. interpolate the error in then - what I was getting at is that we'll want to see the error that came out.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[rpc/context.go, line 167 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGdwjpE22HwF5KGxtI:-KOGwV0y4bS6IdCIT_MJ:677940215) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can this be done eagerly instead?
</details>
ping

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[rpc/context.go, line 167 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGdwjpE22HwF5KGxtI:-KOGwcu8RiXMSLqFVHuW:1901276176) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ping
</details>
Huh, how did I miss that comment? Yeah, it probably could, I was following the pattern of `ctx.conns.cache`. Do you know why that is initialized lazily?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[rpc/context.go, line 167 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGdwjpE22HwF5KGxtI:-KOGxknvvYYC_CiaYVq1:122957367) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea31cac4dac74a9469c0ecf7925112525e50e3f0/rpc/context.go#L167)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Huh, how did I miss that comment? Yeah, it probably could, I was following the pattern of `ctx.conns.cache`. Do you know why that is initialized lazily?
</details>
I've changed both maps to be allocated in `NewContext`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Reviewed 1 of 4 files at r1, 1 of 1 files at r2, 3 of 3 files at r3.
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[rpc/context.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGt6p4K8AzY5MOV_r1:-KOGt6p5mSI4XFTWw90s:460172520) ([raw file](https://github.com/cockroachdb/cockroach/blob/004e10d055b5b743435c2e81e72a9f4b38de0658/rpc/context.go#L174)):*
> ```Go
> 		// intervals between successive attempts to resolve a node address
> 		// and connect via GRPC.
> 		expBO := backoff.NewExponentialBackOff()
> ```

Add a comment (in both breaker definitions?) that increasing MaxInterval beyond the raft election timeout (currently 1.5s) could be disruptive. 

Same comment as https://github.com/cockroachdb/cockroach/pull/8021#issuecomment-236319377:

I think we should set all the major parameters here instead of using the defaults for some of them; it's easier than flipping back to the backoff source and protects us against any upstream changes to the default value.

500ms-1s is a narrow range for an exponential backoff; I think a smaller minimum would make sense (or else a simpler constant backoff).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-:-KOGyKKxoCo7HTcZW27a:676834233)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[rpc/context.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGt6p4K8AzY5MOV_r1:-KOH-GYEEQQJNJIgzb7t:-166821797) ([raw file](https://github.com/cockroachdb/cockroach/blob/004e10d055b5b743435c2e81e72a9f4b38de0658/rpc/context.go#L174)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add a comment (in both breaker definitions?) that increasing MaxInterval beyond the raft election timeout (currently 1.5s) could be disruptive. 
> 
> Same comment as https://github.com/cockroachdb/cockroach/pull/8021#issuecomment-236319377:
> 
> I think we should set all the major parameters here instead of using the defaults for some of them; it's easier than flipping back to the backoff source and protects us against any upstream changes to the default value.
> 
> 500ms-1s is a narrow range for an exponential backoff; I think a smaller minimum would make sense (or else a simpler constant backoff).
</details>
Added `rpc.Context.NewBackoff`. Agreed that 500ms-1s is a narrow range, though I don't have a particular insight into what we should set the backoff to, so I've left as-is for now.

Can you expand on the disruption caused by setting MaxInterval beyond the raft election timeout? I'd like to make the comment accurate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[rpc/context.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGt6p4K8AzY5MOV_r1:-KOH0j8U9GAZNkUCgrLl:788586224) ([raw file](https://github.com/cockroachdb/cockroach/blob/004e10d055b5b743435c2e81e72a9f4b38de0658/rpc/context.go#L174)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Added `rpc.Context.NewBackoff`. Agreed that 500ms-1s is a narrow range, though I don't have a particular insight into what we should set the backoff to, so I've left as-is for now.
> 
> Can you expand on the disruption caused by setting MaxInterval beyond the raft election timeout? I'd like to make the comment accurate.
</details>
I think you didn't push this last update.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[rpc/context.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGt6p4K8AzY5MOV_r1:-KOH12hD-IiydC2yQq8p:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/004e10d055b5b743435c2e81e72a9f4b38de0658/rpc/context.go#L174)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think you didn't push this last update.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[rpc/context.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGt6p4K8AzY5MOV_r1:-KOH3Y0drbF15BSINCAY:1069202650) ([raw file](https://github.com/cockroachdb/cockroach/blob/004e10d055b5b743435c2e81e72a9f4b38de0658/rpc/context.go#L174)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Done.
</details>
A newly restarted node will be in follower mode with no knowledge of the raft leader. If it doesn't hear from a leader before the election timeout expires, it will start to campaign, which can be disruptive. Therefore the leader needs to get in touch with such nodes within one election timeout of their restart, which won't happen if their backoff is too high. 

There may be other alternatives, such as resetting circuit breakers for a node when that node reappears in gossip (assuming gossip propagation time is less than the election timeout), or implementing two-phase elections so that a restarted node campaigning won't be disruptive.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[rpc/context.go, line 221 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOH5X1HLrraEkw_SaWM:-KOH5X1HLrraEkw_SaWN:28086261) ([raw file](https://github.com/cockroachdb/cockroach/blob/51c1ebc49dedf35c46afc156409f5c0b8815b27b/rpc/context.go#L221)):*
> ```Go
> // NewBackOff creates a new exponential backoff properly configured for RPC
> // connection backoff.
> func (ctx *Context) NewBackOff() backoff.BackOff {
> ```

why is this a method? it can be a function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[rpc/context.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOGt6p4K8AzY5MOV_r1:-KOHF5cucJz1eLBRgZc5:735057003) ([raw file](https://github.com/cockroachdb/cockroach/blob/004e10d055b5b743435c2e81e72a9f4b38de0658/rpc/context.go#L174)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> A newly restarted node will be in follower mode with no knowledge of the raft leader. If it doesn't hear from a leader before the election timeout expires, it will start to campaign, which can be disruptive. Therefore the leader needs to get in touch with such nodes within one election timeout of their restart, which won't happen if their backoff is too high. 
> 
> There may be other alternatives, such as resetting circuit breakers for a node when that node reappears in gossip (assuming gossip propagation time is less than the election timeout), or implementing two-phase elections so that a restarted node campaigning won't be disruptive.
</details>
Thanks. Added a comment. Note that I saw exactly that behavior before we fixed the fail-fast setting of heartbeats. A node would come up, wouldn't hear from the leader and would start a campaign. 

---

*[rpc/context.go, line 221 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292#-KOH5X1HLrraEkw_SaWM:-KOHE_bhOvo4gRujx1MK:775633632) ([raw file](https://github.com/cockroachdb/cockroach/blob/51c1ebc49dedf35c46afc156409f5c0b8815b27b/rpc/context.go#L221)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is this a method? it can be a function.
</details>
There is no need for it to be a method. Moved it into a new `breaker.go` file and added a `NewBreaker` function as well. PTAL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8292)*
<!-- Sent from Reviewable.io -->
